src/rascal/models/sparse_kernel_predict.hh

```diff
@@ -0,0 +1,256 @@
+   * @param sparse_points a SparsePoints* class
+   * @param managers a ManagerCollection or similar collection of
+   * structure managers
+   * @param zeta exponent of the GAP kernel
+   * @param weights of the GAP model
+   * @param representation_name name used to get the representation in the
+   * managers
+   * @param representation_grad_name name used to get representation gradient
+   * the in the managers
+   * @param djfi_name name used to get/register the partial gradients in the
+   * managers
+   * @return
+   */
+  template <class Property_t, class PropertyGradient_t, class StructureManager,
+            class SparsePoints>
+  void
+  compute_partial_gradients_gap(StructureManager & manager,
+                                SparsePoints & sparse_points,
+                                math::Vector_t & weights, const size_t zeta,
+                                const std::string & representation_name,
+                                const std::string & representation_grad_name,
+                                const std::string & djfi_name) {
+    using Manager_t = typename StructureManager::element_type;
+    using Keys_t = typename SparsePoints::Keys_t;
+    using Key_t = typename SparsePoints::Key_t;
+
+    auto && prop{
+        *manager->template get_property<Property_t>(representation_name, true)};
+    auto && prop_grad{*manager->template get_property<PropertyGradient_t>(
+        representation_grad_name, true)};
+
+    const int inner_size{prop.get_nb_comp()};
+
+    bool do_block_by_key_dot{false};
+    if (prop_grad.are_keys_uniform()) {
+      do_block_by_key_dot = true;
+    }
+
+    std::set<int> unique_species{};
+    for (auto center : manager) {
+      unique_species.insert(center.get_atom_type());
+    }
+
+    // find shared central atom species
+    std::set<int> species_intersect{
+        internal::set_intersection(unique_species, sparse_points.species())};
+
+    // find offsets alongs the sparse points direction
+    std::map<int, int> offsets{sparse_points.get_offsets()};
+    Keys_t rep_keys{prop_grad.get_keys()};
+    std::map<int, Keys_t> keys_intersect{};
+    for (const int & sp : species_intersect) {
+      keys_intersect[sp] =
+          internal::set_intersection(rep_keys, sparse_points.keys_sp.at(sp));
+    }
+
+    // attach partial gradients array to manager
+    auto && djfi{
+        *manager
+             ->template get_property<Property<double, 2, Manager_t, 1, ThreeD>>(
+                 djfi_name, true, true)};
+    // don't recompute the partial gradients if already up to date
+    if (djfi.is_updated()) {
+      return;
+    }
+
+    djfi.resize();
+    djfi.setZero();
+
+    if (species_intersect.empty()) {
+      return;
+    }
+
+    math::Vector_t weights_scaled(weights.size());
+    size_t i_row{0};
+    for (auto center : manager) {
+      const int a_sp{center.get_atom_type()};
+      // \alpha_n^{scaled} = \alpha_n * [z* (X_j \dot T_n)^{z-1}]
```

Comment:
In words: This is what comes out of the derivative of the kernel when the chain rule is applied.

Reply:
there is the models weights so not quite I would say.

Reply:
...so model weights times chain-rule detritus? It's much easier to understand conceptually what's being done here when you write it in words (it's easier than you think) rather than equations, which anyway kinda just repeat the line below.

Reply:
I personally prefer the equation as it tells me which operation is being done precisely while a more conceptual comment is not self sufficient to understand the following line. The relation between the code and the equation is far from direct to me.

Reply:
And that's precisely why you need both a plain-text explanation (for when we come back to this in six months and don't remember what the symbols mean) and the equation, which doesn't necessarily have to be in an inline comment. We could, for instance, put the different steps in a LaTeX document in the developer documentation, like we've done with the spherical expansion derivation (I think this needs updating btw...) -- it's just that inline comments are horrible to read for those of us without a LaTeX compiler in our heads.